OpenKeychain/src/main/java/org/sufficientlysecure/keychain/keyimport/ImportKeysList.java

```diff
@@ -78,13 +79,19 @@ private boolean mergeDupes(ImportKeysListEntry incoming, ImportKeysListEntry exi
             // to work properly, Keybase-sourced entries need to pass along the extra
             if (KeybaseKeyserver.ORIGIN.equals(origin)) {
                 existing.setExtraData(incoming.getExtraData());
+                incomingFromHkpServer = false;
             }
         }
+
         ArrayList<String> incomingIDs = incoming.getUserIds();
         ArrayList<String> existingIDs = existing.getUserIds();
         for (String incomingID : incomingIDs) {
             if (!existingIDs.contains(incomingID)) {
-                existingIDs.add(incomingID);
+                if (incomingFromHkpServer) {
```

Comment:
User IDs / email addresses returned from HKP servers are preferred over those from other (keybase.io) servers.
This prepends HKP server results to the start of the list, so that the UI (for cloud key search), picking the first list element, shows the right email address.